### PR TITLE
Renode toolset

### PR DESCRIPTION
**Pull request by IshDeshpa**
Grabbed some of the stuff from renode-test-framework and put it into this branch. Currently includes helpful scripts and VSCode integrations for testing with renode. Also renamed many files which is why this touches so many things.

### CODE REVIEW

File: Scripts/run_test.sh [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+source ${HOME}/.bashrc
```
Comment (Cam0Cow) [inline Scripts/run_test.sh:2]: Why is this needed?

File: Tests/Leader/Test_Template.c [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        (OS_TICK)NULL,
+        (void *)NULL,
+        (OS_OPT)(OS_OPT_TASK_STK_CLR),
+        (OS_ERR *)&err);
+    assertOSError(OS_MAIN_LOC, err);
+}
+
+
+static void StartTask(void *arg){
+    // Spawn any required applications here.
```
Comment (Cam0Cow) [inline Tests/Leader/Test_Template.c:48]: Should probably add systick initialization here, you always need it

File: Scripts/load_env_vars.sh [hunk 3]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+export CONTROLS_REPO=$(dirname $(dirname $(realpath "$BASH_SOURCE")))
```
Comment (Cam0Cow) [inline Scripts/load_env_vars.sh:3]: This seems finicky and will break if the script ever gets moved into a subdirectory. Maybe change it to `export CONTROLS_REPO=$(git -C $BASH_SOURCE rev-parse --show-toplevel)` which will work as long as the script is in the repo

**Review by Cam0Cow — CHANGES_REQUESTED** [review]

File: Scripts/run_test.sh (hunk 1 above)
Comment (IshDeshpa) [reply Scripts/run_test.sh:2]: Technically this will be run whenever a terminal is started up. I believe it was added for testing whether the environment variables were actually set. I will remove it.

**Comment by IshDeshpa** [comment]
Instructions on how to test:
- Open the LHR code workspace in .vscode from VSCode
- Try running some of the tasks by pressing Ctrl-Shift-P and hitting Tasks: Run Task
- Try debugging via the Run and Debug option on the left sidebar. This should attach directly to gdb through the Debug Console on the bottom panel, as well as open up Renode so that you can start the simulation. You can add breakpoints in the code (see Test_HelloWorld) directly through VSCode.

**Review by Cam0Cow — DISMISSED** [review]
Works for debugging on Renode

File: .gitignore [hunk 4]
```diff
@@ -4,3 +4,7 @@ __pycache__
 *.csv
 .vscode
 *.swp
+Tests/Results
+
+!.vscode/LHR.code-workspace
+!.vscode/c_cpp_properties.json
```
Comment (NathanielDelgado) [inline .gitignore:10]: Just ignore the folder

File: BSP/STM32F413/Makefile [hunk 5]
```diff
@@ -111,8 +110,8 @@ C_INCLUDES =  \
 -I../../RTOS/uCOS-III-STM32F4/uCOS-III/Ports/ARM-Cortex-M4/Generic/GNU/ \
 -I../../RTOS/uCOS-III-STM32F4/uC-CPU/ \
 -I../../RTOS/uCOS-III-STM32F4/uC-CPU/ARM-Cortex-M4/GNU/ \
--I../../RTOS/uCOS-III-STM32F4/uC-LIB/
-
+-I../../RTOS/uCOS-III-STM32F4/uC-LIB/ \
+-I../../Tests/Inc/ \
```
Comment (NathanielDelgado) [inline BSP/STM32F413/Makefile:114]: Why would tests need an include folder?

File: Makefile [hunk 6]
```diff
@@ -10,13 +10,33 @@ DARKGRAY=\033[1;30m
...
 export DEBUG
 
+# Note: ?= will define the variable only if it is not already set, which means the user can override any of
+# the TEST_LEADER, TEST_MOTOR, or TEST_CAR assignments with their own test files.
+
+# Check if test file exists for the leader.
+ifneq (,$(wildcard Tests/Leader/Test_$(TEST).c))
+	TEST_LEADER ?= Tests/Leader/Test_$(TEST).c
+else
+	TEST_LEADER ?= Apps/Src/main.c
```
Comment (NathanielDelgado) [inline Makefile:23]: Shouldn't we just give them a warning

File: Renode/STM32DMA_Fix.cs
Comment (NathanielDelgado) [inline Renode/STM32DMA_Fix.cs:1]: I'm assuming this is okay

**Review by NathanielDelgado — APPROVED** [review]
I didn't check the .cs files
Looks okay

File: .gitignore (hunk 4 above)
Comment (Cam0Cow) [reply .gitignore:10]: The exclamation marks are a whitelist. We specifically care about those files.

File: Makefile (hunk 6 above)
Comment (Cam0Cow) [reply Makefile:23]: Nah, it just means the default is using main instead of a test.

**Review by KnockbackNemo — DISMISSED** [review]
I didn't look through the files, but when I used it for testing, everything worked great for me!

**Review by Cam0Cow — APPROVED** [review]

File: .gitignore [hunk 7]
```diff
@@ -4,3 +4,7 @@ __pycache__
 *.csv
 .vscode
 *.swp
+Tests/Results
```
Comment (NathanielDelgado) [inline .gitignore:7]: TAKE THIS OUT!!!

File: .vscode/LHR.code-workspace [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+                    "traceResponse": false
+                 }
+            }
+
+        ]
+    },
+    "extensions": {
+        "recommendations": [
+            "ms-vscode.cpptools-extension-pack",    // For Controls C code
+            "GitHub.vscode-pull-request-github",    // For reviewing pull requests directly from VSCode
```
Comment (NathanielDelgado) [inline .vscode/LHR.code-workspace:291]: TAKE THIS OUT

File: Apps/Inc/Tasks.h [hunk 9]
```diff
@@ -180,7 +180,10 @@ void throwTaskError(error_code_t errorCode, callback_t errorCallback, error_sche
...
 
 #if DEBUG == 1
-#define assertOSError(err) 
+#define assertOSError(err) \
+        if (err != OS_ERR_NONE) { \
+            printf("Error asserted at %s, line %d: %d\n\r", __FILE__, __LINE__, err); \
+        } \
         _assertOSError(err);
 #else
 #define assertOSError(err) _assertOSError(err);
```
Comment (NathanielDelgado) [inline Apps/Inc/Tasks.h:189]: WOULDN'T IT MAKE MORE SENSE TO HAVE THIS IN THE _ASSWERTOSFUNCTION ????

File: BSP/STM32F413/Makefile [hunk 10]
```diff
@@ -8,7 +8,11 @@ TARGET = controls-leader
 # building variables
 ######################################
 # optimization
-OPT = -Og -g
+ifeq ($(DEBUG), 0)
+OPT = -O3
+else
+OPT = -Og
```
Comment (NathanielDelgado) [inline BSP/STM32F413/Makefile:14]: ADD -G

File: BSP/STM32F413/Makefile [hunk 11]
```diff
@@ -32,15 +36,10 @@ $(wildcard ../../RTOS/uCOS-III-STM32F4/uC-CPU/*.c) \
...
-TEST_FILE := $(TEST).c
-# This line adds everything in Apps/Src/*.c except for main.c
+# This line adds everything in Apps/Src/*.c except for main.c, then adds the test file
 C_SOURCES += \
 $(filter-out ../../Apps/Src/main.c, $(wildcard ../../Apps/Src/*.c))	\
-$(wildcard ../../$(TEST_FILE))
-else
-C_SOURCES += $(wildcard ../../Apps/Src/*.c)
-endif
+../../$(TEST)
```
Comment (NathanielDelgado) [inline BSP/STM32F413/Makefile:42]: where's the beef?

File: Makefile [hunk 12]
```diff
@@ -62,7 +41,7 @@ docs:
 help:
 	@echo "Format: ${ORANGE}make ${BLUE}<BSP type>${NC}${ORANGE}TEST=${PURPLE}<Test type>${NC}"
 	@echo "BSP types (required):"
-	@echo "	${BLUE}simulator${NC}"
+	@echo " BSP types (required):"
 	@echo "	${BLUE}stm32f413${NC}"
 	@echo ""
 	@echo "Test types (optional):"
```
Comment (NathanielDelgado) [inline Makefile:47]: update his

File: Renode/startup.resc [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+macro show_uart_leader
+"""
+    emulation CreateServerSocketTerminal $uart2_telnet "ldr-uart-con" false
+    connector Connect sysbus.usart2 ldr-uart-con
+"""
+
+macro show_uart_display
+"""
+    emulation CreateServerSocketTerminal $uart3_telnet "disp-uart-con" false
+    connector Connect sysbus.usart3 disp-uart-con
```
Comment (NathanielDelgado) [inline Renode/startup.resc:27]: ### _**ADD STRINGS TO DISPLAY WHAT PORTS THEY ARE DISPLAYED ON**_

File: Tests/Test_BSP_CAN.c
Comment (NathanielDelgado) [inline Tests/Test_BSP_CAN.c:1]: DELETE!???? 


???

File: Tests/Test_HelloWorld.c
Comment (NathanielDelgado) [inline Tests/Test_HelloWorld.c:1]: love this.

**Review by NathanielDelgado — CHANGES_REQUESTED** [review]
comments

File: Apps/Inc/Tasks.h (hunk 9 above)
Comment (IshDeshpa) [reply Apps/Inc/Tasks.h:189]: Can't be in the assert OS function bc otherwise __FILE__ and __LINE__ would just show Tasks.h and line 185.

File: .vscode/LHR.code-workspace (hunk 8 above)
Comment (IshDeshpa) [reply .vscode/LHR.code-workspace:291]: Why?

File: Renode/startup.resc (hunk 13 above)
Comment (IshDeshpa) [reply Renode/startup.resc:27]: too hard.

**Review by Cam0Cow — APPROVED** [review]

**Review by Cam0Cow — CHANGES_REQUESTED** [review]
Delete this

File: Apps/Inc/Tasks.h (hunk 9 above)
Comment (NathanielDelgado) [reply Apps/Inc/Tasks.h:189]: bruh

File: Simulator/bpsSim.elf
Comment (NathanielDelgado) [inline Simulator/bpsSim.elf:1]: ♘💜  тคкＥ тℍＩ𝐒 ㄖＵｔ!!!

!!!  🍔🐍

File: Tests/Test_HelloWorld.c
Comment (NathanielDelgado) [inline Tests/Test_HelloWorld.c:1]: Yes

**Review by NathanielDelgado — APPROVED** [review]
thanks for tihs

**Review by Cam0Cow — APPROVED** [review]
85q03f
